File: lintcode/permutation_sequence.py
class Solution:
    """
    @param n: n
    @param k: the k-th permutation
    @return: a string, the k-th permutation
    """
    def getPermutation(self, n, k):
        fac, result = 1, [i + 1 for i in xrange(n)]
        for i in xrange(2, n):
            fac *= i
        k -= 1
        for i in xrange(n - 1):
            (d, k) = divmod(k, fac)
            if d: result[i:i + d + 1] = [result[i + d]] + result[i:i + d]
            fac /= (n - 1 - i)
        return ''.join([str(x) for x in result])

File: leetcode/permutation_sequence.js
/* Given n and k, return the kth permutation sequence.
 * Time Complexity: O(n^2)
 * Space Complexity: O(n), assuming rotation in place
 */

/**
 * @param {number} n
 * @param {number} k
 * @return {string}
 */
var getPermutation = function(n, k) {
    var result = [], fac = 1;
    for (var i = 0; i < n; ++i) {
        result.push(i + 1);
        if (i > 1)
            fac *= i;
    }
    for (--k, i = 0; i < n - 1; ++i) {
        var d = Math.floor(k / fac);
        k -= fac * d;
        if (d > 0) {
            var tmp = result[i + d];
            for (var j = i + d; j > i; --j)
                result[j] = result[j - 1];
            result[i] = tmp;
        }
        fac = Math.floor(fac / (n - 1 - i));
    }
    return result.join("");
};
